model_compression_toolkit/data_generation/common/optimization_utils.py

```diff
@@ -0,0 +1,460 @@
+            bn_alignment_loss_fn (Callable): the batch norm alignment loss function.
+            layer_weights (Dict): weights to multiply the loss for each layer.
+
+        Returns:
+            Any: the computed batch norm alignment loss.
+        """
+        # Update the batch statistics for the current batch
+        self.all_imgs_stats_holder.update_batch_stats(batch_index=batch_index,
+                                                      input_imgs=input_imgs,
+                                                      activation_extractor=activation_extractor,
+                                                      to_differentiate=True)
+
+        # Initialize variables for accumulating the batchnorm alignment loss
+        total_bn_loss = 0
+
+        # Iterate over each BN layer
+        for layer_name in orig_bn_stats_holder.get_bn_layer_names():
+            # Get the layer weight for the current BN layer
+            layer_weight = layer_weights.get(layer_name)
+
+            # Get the mean and variance from the original BN statistics
+            bn_layer_mean = orig_bn_stats_holder.get_mean(layer_name)
+            bn_layer_std = orig_bn_stats_holder.get_std(layer_name)
+
+            # Get the mean and variance from the current batch's statistics
+            if self.use_all_data_stats:
+                # If using all data statistics, retrieve the accumulated statistics from all batches
+                imgs_layer_mean, imgs_layer_var, imgs_layer_std = self.get_accumulated_stats_per_layer(layer_name)
+            else:
+                # Otherwise, retrieve the statistics from the current batch
+                imgs_layer_mean, imgs_layer_second_moment, imgs_layer_var, imgs_layer_std = self.all_imgs_stats_holder.get_stats(batch_index, layer_name)
+
+            # Accumulate the batchnorm alignment weighted by the layer weight
+            total_bn_loss += layer_weight * bn_alignment_loss_fn(bn_layer_mean, imgs_layer_mean, bn_layer_std, imgs_layer_std)
+
+        return total_bn_loss
+
+    def update_statistics(self,
+                        input_imgs: Any,
+                        batch_index: int,
+                        activation_extractor: ActivationExtractor):
+        """
+        Update the statistics for the images at the specified batch index.
+
+        Args:
+            input_imgs (Any): the input images.
+            batch_index (int): the index of the batch.
+            activation_extractor (ActivationExtractor): extractor for layer activations.
+        """
+        if self.use_all_data_stats:
+            self.all_imgs_stats_holder.update_batch_stats(batch_index=batch_index,
+                                                          input_imgs=input_imgs,
+                                                          activation_extractor=activation_extractor,
+                                                          to_differentiate=False)
+
+    def optimization_step(self,
```

Comment:
declare as an abstract method to enforce implementation? (same for other methods below)

Reply:
Done